virtual-desktop/src/app/start-url-manager/start-url-manager.service.ts

```diff
@@ -0,0 +1,141 @@
+  private readonly logger: ZLUX.ComponentLogger = BaseLogger;
+
+  constructor(
+    public parser: App2AppArgsParser,
+  ) {
+    this.registerTestAction();
+  }
+
+  onLogin(_username: string, _plugins: ZLUX.Plugin[]): boolean {
+    if (this.done) {
+      return true;
+    }
+    this.done = true;
+    const allArgs = this.getAllApp2AppArgsFromURL();
+    if (!allArgs) {
+      return true;
+    }
+    for (const args of allArgs) {
+      if (this.validateApp2AppArgs(args)) {
+        this.invokeApp2App(args);
+      }
+    }
+    return true;
+  }
+
+  private validateApp2AppArgs(args: App2AppArgs): boolean {
+    const { dispatcher, pluginManager } = ZoweZLUX;
+    const plugin = pluginManager.getPlugin(args.pluginId);
+    if (!plugin) {
+      // this.logger.warn('Plugin not found for identifier %s specified in app2app query URL parameter', args.pluginId);
+      this.logger.warn('ZWED5193W', args.pluginId);
+      return false;
+    }
+    if (args.actionType !== 'launch' && args.actionType !== 'message') {
+      // this.logger.warn('Action Type(%s) specified  in app2app query URL parameter is invalid', args.actionType);
+      this.logger.warn('ZWED5194W', args.actionType);
+      return false;
+    }
+    if (args.actionMode !== 'create' && args.actionMode !== 'system') {
+      // this.logger.warn('Action Mode(%s) specified  in app2app query URL parameter is invalid', args.actionMode);
+      this.logger.warn('ZWED5195W', args.actionMode);
+      return false;
+    }
+    try {
+      JSON.parse(args.contextData);
+    } catch (e) {
+      // this.logger.warn('Context Data(%s) specified in app2app query URL parameter is not valid JSON', args.contextData);
+      this.logger.warn('ZWED5196W', args.contextData);
+      return false;
+    }
+    if (args.formatter !== 'data') {
+      const abstractAction = dispatcher.getAbstractActionById(args.formatter);
+      if (!abstractAction) {
+        // this.logger.warn('Unable to find registered action for identifier %s specified in app2app query URL parameter', args.formatter);
+        this.logger.warn('ZWED5197W', args.formatter);
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private invokeApp2App(args: App2AppArgs): void {
+    const { dispatcher } = ZoweZLUX;
+    const ActionType = dispatcher.constants.ActionType;
+    const ActionTargetMode = dispatcher.constants.ActionTargetMode;
+    const targetPluginId = args.pluginId;
+    const type = args.actionType === 'launch' ? ActionType.Launch : ActionType.Message;
+    const mode = args.actionMode === 'create' ? ActionTargetMode.PluginCreate : ActionTargetMode.System;
+    let contextData: any = JSON.parse(args.contextData);
+    let action: ZLUX.Action;
+    let argumentData: any;
+    if (args.formatter === 'data') {
+      const actionTitle = 'Launch app from URL';
+      const actionId = 'org.zowe.zlux.url.launch';
+      const argumentFormatter = { data: { op: 'deref', source: 'event', path: ['data'] } };
```

Comment:
I think we could bake this into zlux-platform dispatcher, so it makes it at startup. That way, we could call `dispatcher.getAbstractActionById('data');`
The reason why is because I am not sure what happens to the dispatcher if we do multiple `makeAction` calls for the same name

Reply:
I took a look at `https://github.com/zowe/zlux-platform/blob/master/base/src/dispatcher/dispatcher.ts#L679` and deeper. `dispatcher.makeAction()` doesn't register the action that it has created.